selectel/iam.go

```diff
@@ -0,0 +1,196 @@
+		}
+	}
+
+	for _, newRole := range newRoles {
+		if !slices.Contains(oldRoles, newRole) {
+			rolesToAssign = append(rolesToAssign, newRole)
+		}
+	}
+
+	return rolesToUnassign, rolesToAssign
+}
+
+func applyServiceUserRoles(ctx context.Context, d *schema.ResourceData, iamClient *iam.Client, rolesToUnassign, rolesToAssign []roles.Role) error {
+	if len(rolesToAssign) != 0 {
+		err := iamClient.ServiceUsers.AssignRoles(ctx, d.Id(), rolesToAssign)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(rolesToUnassign) != 0 {
+		err := iamClient.ServiceUsers.UnassignRoles(ctx, d.Id(), rolesToUnassign)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func applyUserRoles(ctx context.Context, d *schema.ResourceData, iamClient *iam.Client, rolesToUnassign, rolesToAssign []roles.Role) error {
+	if len(rolesToAssign) != 0 {
+		err := iamClient.Users.AssignRoles(ctx, d.Id(), rolesToAssign)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(rolesToUnassign) != 0 {
+		err := iamClient.Users.UnassignRoles(ctx, d.Id(), rolesToUnassign)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func convertIAMListToUserFederation(federationList []interface{}) (*users.Federation, error) {
+	if len(federationList) == 0 {
+		return nil, nil
+	}
+	if len(federationList) > 1 {
+		return nil, errors.New("more than one federation value provided")
+	}
+	var idRaw, externalIDRaw interface{}
+	var ok bool
+
+	if idRaw, ok = federationList[0].(map[string]interface{})["id"]; !ok {
+		return nil, errors.New("federation.id value isn't provided")
+	}
+	if externalIDRaw, ok = federationList[0].(map[string]interface{})["external_id"]; !ok {
+		return nil, errors.New("federation.external_id value isn't provided")
+	}
+
+	id := idRaw.(string)
+	externalID := externalIDRaw.(string)
+
+	federation := &users.Federation{
+		ExternalID: externalID,
+		ID:         id,
+	}
+
+	return federation, nil
+}
+
+func convertIAMSetToRoles(rolesSet *schema.Set) ([]roles.Role, error) {
+	rolesList := rolesSet.List()
+
+	output := make([]roles.Role, len(rolesList))
+	var roleNameRaw, scopeRaw, projectIDRaw interface{}
+	var ok bool
+
+	for i := range rolesList {
+		var roleName, scope, projectID string
+		resourceRoleMap := rolesList[i].(map[string]interface{})
+
+		if roleNameRaw, ok = resourceRoleMap["role_name"]; !ok {
+			return nil, errors.New("role_name value isn't provided")
+		}
+		if scopeRaw, ok = resourceRoleMap["scope"]; !ok {
+			return nil, errors.New("scope value isn't provided")
+		}
+		if projectIDRaw, ok = resourceRoleMap["project_id"]; ok {
+			projectID = projectIDRaw.(string)
+		}
+
+		roleName = roleNameRaw.(string)
+		scope = scopeRaw.(string)
+
+		output[i] = roles.Role{
+			RoleName:  roles.Name(roleName),
+			Scope:     roles.Scope(scope),
+			ProjectID: projectID,
+		}
+	}
+
+	return output, nil
+}
+
+func convertIAMRolesToSet(roles []roles.Role) []interface{} {
+	result := make([]interface{}, 0)
```

Comment:
nit:
здесь нам известна длина итогового  слайса, так что можно сразу создать с нужным capacity
```suggestion
	result := make([]interface{}, 0, len(roles))
```

Reply:
Да кстати, спасибо, исправил!